Change Configuration Documentation

eljobe opened the pull request:
### Context
We need to do a better job explaining how configurations work, and specifically to encourage users to migrate from the Java Plugin to the Java Library Plugin as appropriate.

See background at: #1854

eljobe commented:
@oehme and @adammurdoch, I'm really not that crazy about the changes I have added in this Pull Request, but it's hard to figure out exactly the right amount of detail to explain at this point in our development of these newer configurations.

One thing I've considered, but not yet done is the idea of adding another graphic to show the differences between the old and new configuration hierarchies. This graphic would be specifically targeting users who are looking for an easy migration path.

Please advise me on how to proceed.

oehme commented on subprojects/docs/src/docs/userguide/javaPlugin.xml:
```diff
@@ -378,7 +378,10 @@
                 <td>-</td>
-                <td>Compile time dependencies</td>
+                <td>
+                    Compile time dependencies. This configuration is weakly deprecated in favor of the <code>implementation</code> configuration defined
```
❌ We should add `implementation`, `testImplementation` and `runtimeOnly` to this list instead of linking to the java-library plugin. These configurations are available in the `java` plugin already. Instead, we can either link back here from `java-library` or just duplicate it in both places.

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/javaPlugin.xml:
```diff
@@ -396,7 +399,10 @@
                 <td>compile</td>
                 <td>-</td>
-                <td>Runtime dependencies</td>
+                <td>Runtime dependencies. This configuration is weakly deprecated in favor of the <code>implementation</code>, <code>runtimeElements</code>
```
❌ `runtimeOnly` is the replacement for `runtime`, not `implementation`. `implementation is visible at compile time and is one of the replacements for `compile` (the other being `api` in the `java-library` plugin.

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/javaPlugin.xml:
```diff
@@ -420,7 +426,9 @@
                 <td>runtime, testCompile</td>
                 <td>test</td>
-                <td>Additional dependencies for running tests only</td>
+                <td>Additional dependencies for running tests only. This configuration is weakly deprecated in favor of the <code>testImplementation</code> or <code>testRuntimeClasspath</code>
```
`testRuntimeOnly`, same as above above.

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml:
```diff
@@ -71,44 +71,68 @@
+                        The dependencies required by the production classes at runtime. By default, this includes the
+                        dependencies declared in the <code>api</code>, <code>implementation</code>, and <code>runtimeOnly</code> configurations.
+                        This configuration is an example of a configuration used for Resolving Dependencies, and as such, it should be unusual for
+                        users to declare dependencies directly in the <code>runtimeClasspath</code> configuration.
```
They should *never* add dependencies to `runtimeClasspath`.

oehme commented on subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml:
```diff
@@ -71,44 +71,68 @@
-        <para>
-            The Java plugin defines a number of standard configurations. These configurations represent the classpaths that the
-            Java plugin uses. Some are listed below, and you can find more details in <xref linkend="tab:configurations"/>.
+        <para>In Gradle dependencies are grouped into <firstterm>configurations</firstterm>. A configuration is a named set of dependencies.
```
I'd shorten this to just 

"A Configuration is a named set of dependencies and artifacts. There are three main purposes for a Configuration:"

eljobe replied:
Done

eljobe commented:
I think I addressed your comments @oehme, but please take another look. I'm also not sure I updated the right image. But, I think I did.

One thing I find confusing about all of this is that some of the configurations are put in place by the Java Plugin while others (just `api` and `apiElements`?) are added by  the Java Library Plugin. It's hard, for example, to explain on the Java Plugin page that the `implementation` configuration extends the `api` configuration when we don't really mention the `api` configuration in the rest of the document.

oehme commented:
> t's hard, for example, to explain on the Java Plugin page that the implementation configuration extends the api configuration when we don't really mention the api configuration in the rest of the document.

We shouldn't mention that `extends` relationship in the `java` plugin guide, because `api` doesn't exist in the `java` plugin. We should just tell users "if your project is a library and needs to expose api dependencies to clients, use the `java-library` plugin.

oehme commented on subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml:
```diff
@@ -71,44 +71,61 @@
+                <varlistentry>
+                    <term>Exposing Artifacts for Consumption</term>
+                    <listitem>
+                        <para>The plugin uses configurations to define what <firstterm>artifacts</firstterm> it generates for other projects or subprojects to consume.</para>
```
There is no difference between "projects" and "subprojects

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml:
```diff
@@ -71,44 +71,61 @@
+                <varlistentry>
+                    <term>Resolving Dependencies</term>
+                    <listitem>
+                        <para>The plugin uses configurations to find (and possibly download) the files it needs from outside the project to perform the work of its tasks,
```
I find this sentence hard to follow, especially the "coordinate and reason" part. The essence should be "These are used as task inputs".

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/artifactDependenciesTutorial.xml:
```diff
@@ -71,44 +71,61 @@
                     <para>
-                        The dependencies required to run the tests. By default, also includes the compile, runtime and test compile
-                        dependencies.
+                        The external dependencies which are part of this project's externally consumable API as well as the classes which are defined in this project which should be
```
remove "external"

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/javaPlugin.xml:
```diff
@@ -375,10 +375,21 @@
-                <td>Compile time dependencies</td>
+                <td>
+                    Compile time dependencies. This configuration is weakly deprecated in favor of the <code>implementation</code> configuration defined
+                    in <xref linkend="sec:java_library_configurations_graph"/>
```
We should remove the reference to the java library plugin, `implementation` is described just a few lines above.

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/javaPlugin.xml:
```diff
@@ -440,7 +472,7 @@
             <title>Java plugin - dependency configurations</title>
             <imageobject>
-                <imagedata fileref="img/javaPluginConfigurations.png"/>
+                <imagedata fileref="img/java-library-ignore-deprecated-main.png"/>
```
We can't reuse the `java-library` graph 1:1 here, because it contains the `api` configuration, which the `java` plugin doesn't have.

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/javaPlugin.xml:
```diff
@@ -375,10 +375,21 @@
             </thead>
             <tr>
+                <td>implementation</td>
+                <td>api</td>
```
This is not the case. This relationship is only added by the `java-library` plugin and should thus be explained there.

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/javaPlugin.xml:
```diff
@@ -361,8 +361,8 @@
 
```
I'm missing a general pointer towards the `java-library` plugin, which is what users will want to use when their project is a library (which is the 95% case).

eljobe replied:
Added to the end of the first paragraph.

oehme commented on subprojects/docs/src/docs/userguide/img/java-configurations.dot:
```diff
@@ -0,0 +1,21 @@
+digraph configurations {
```
We should show `apiElements` as well as the old `compile` and `runtime` configurations here, so people know where the soft-deprecated ones fit in.

eljobe replied:
Done

oehme commented on subprojects/docs/src/docs/userguide/img/java-configurations.dot:
```diff
@@ -0,0 +1,21 @@
+   node [fontname = "Sans"];
+   edge [fontname = "Sans"];
+
+   "compileClasspath" [shape="box", label=<<B>compileClasspath</B>(R)>, color="lightblue", bgcolor="lightblue", style="filled"]
```
The C/R labels aren't explained anywhere, I'd probably remove them.

eljobe replied:
Done
